File: go/raw_test.go
package mailack

import (
	"bytes"
	"context"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestRawDownloads(t *testing.T) {
	payload := []byte{0, 255, 13, 10, 128}
	for _, status := range []int{200, 404} {
		for _, event := range []bool{false, true} {
			path, header := "/v1/messages/m/raw", "X-Mailack-Canonical-Hash"
			if event {
				path, header = "/v1/messages/m/events/e/raw", "X-Mailack-Raw-SHA256"
			}
			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				if r.Method != "GET" || r.URL.Path != path || r.Header.Get("Authorization") != "Bearer test" {
					t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
				}
				w.Header().Set(header, "digest")
				w.WriteHeader(status)
				if status == 404 {
					w.Write([]byte(`{"error":{"code":"not_found"}}`))
					return
				}
				w.Write(payload)
			}))
			client := NewClient(server.URL, WithAPIKey("test"))
			var data []byte
			var hash string
			var err error
			if event {
				result, e := client.GetEventRaw(context.Background(), "m", "e")
				err = e
				if result != nil {
					data, hash = result.Data, result.RawSHA256
				}
			} else {
				result, e := client.GetMessageRaw(context.Background(), "m")
				err = e
				if result != nil {
					data, hash = result.Data, result.CanonicalHash
				}
			}
			server.Close()
			if status == 404 {
				var apiErr *APIError
				if !errors.As(err, &apiErr) || apiErr.Code != "not_found" {
					t.Fatalf("unexpected error: %v", err)
				}
			} else if err != nil || !bytes.Equal(data, payload) || hash != "digest" {
				t.Fatalf("data=%v hash=%s err=%v", data, hash, err)
			}
		}
	}
}

File: javascript/raw.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { Client, APIError } from './mailack.js';

for (const [method, path, field, header] of [
  ['getMessageRaw', '/v1/messages/m/raw', 'canonicalHash', 'x-mailack-canonical-hash'],
  ['getEventRaw', '/v1/messages/m/events/e/raw', 'rawSha256', 'x-mailack-raw-sha256'],
]) {
  test(method, async () => {
    const bytes = new Uint8Array([0, 255, 13, 10, 128]);
    const client = new Client({ baseUrl: 'https://example.test', apiKey: 'test', fetch: async (url, opts) => {
      assert.equal(url, 'https://example.test' + path);
      assert.equal(opts.method, 'GET');
      assert.equal(opts.headers.Authorization, 'Bearer test');
      return new Response(bytes, { headers: { [header]: 'digest' } });
    }});
    const result = await client[method]('m', 'e');
    assert.deepEqual(new Uint8Array(result.data), bytes);
    assert.equal(result[field], 'digest');
  });
  test(method + ' errors', async () => {
    const client = new Client({ baseUrl: 'https://example.test', fetch: async () =>
      new Response(JSON.stringify({error: {code: 'not_found', message: 'missing'}}), {status: 404}) });
    await assert.rejects(client[method]('m', 'e'), e => e instanceof APIError && e.code === 'not_found');
  });
}

for (const [method, verb, suffix] of [
  ['sealMessage', 'POST', 'seal'],
  ['getEvidence', 'GET', 'evidence'],
  ['getProofBundle', 'GET', 'proof-bundle'],
]) {
  test(method + ' regression', async () => {
    const expected = { canonical_hash: 'digest', proof_path: [] };
    const client = new Client({ baseUrl: 'https://example.test', fetch: async (url, opts) => {
      assert.equal(url, `https://example.test/v1/messages/m/${suffix}`);
      assert.equal(opts.method, verb);
      return Response.json(expected);
    }});
    assert.deepEqual(await client[method]('m'), expected);
  });
}
